### PR TITLE
feat(dht): Check version compatibility on handshakes

### PR DESCRIPTION
**Pull request by juslesan**
## Summary

Check version compatibility on handshakes.

This ensures that nodes running incompatible older versions of the software cannot join networks that assume the use of newer versions.

## Future Improvements

- The version field might no longer be needed in the handshake response in the future. Once all nodes add the version number to the handshake request we can safely assume that the requests version is enough to decide compatibility.
- The version numbering could be separated from the monorepo version to a protocol version. The Protocol version could be sent instead of the monorepo version. This would ensure that backwards breaking changes to the client or broker would not require network splitting on the lower layers.
- The fallback value `BEFORE_TESTNET_TWO_VERSION` will likely not be needed in the future and can be removed.

### CODE REVIEW

File: packages/dht/src/helpers/versionCompatibility.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    const minorSourceVersion = excludePatchVersion(version1)
+    const minorOwnVersion = excludePatchVersion(version2)
+    return minorSourceVersion === minorOwnVersion
+}
+
+export const excludePatchVersion = (version: string): string => {
+    const versionParts = version.split('.')
+    versionParts.pop()
+    return versionParts.join('.')
+}
```
Comment (harbu) [inline packages/dht/src/helpers/versionCompatibility.ts:13]: Maybe a unit test case for empty string ensuring it doesn't throw.

File: packages/dht/src/helpers/versionCompatibility.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+// Is able to compare versions such as 1.2.3 and 1.2.4
+// can also compare versions such as 100.0.0-pretestnet.0 and 100.0.0-pretestnet.40
+export const isCompatibleVersion = (version1: string, version2: string): boolean => {
+    const minorSourceVersion = excludePatchVersion(version1)
+    const minorOwnVersion = excludePatchVersion(version2)
+    return minorSourceVersion === minorOwnVersion
```
Comment (teogeb) [inline packages/dht/src/helpers/versionCompatibility.ts]:
```suggestion
    const minorVersion1 = excludePatchVersion(version1)
    const minorVersion2 = excludePatchVersion(version2)
    return minorVersion1 === minorVersion2
```

File: packages/dht/src/connection/websocket/WebsocketConnector.ts [hunk 3]
```diff
@@ -308,7 +315,9 @@ export class WebsocketConnector {
                 targetPeerDescriptor
             )
             managedConnection.setRemotePeerDescriptor(sourcePeerDescriptor)
-            if (targetPeerDescriptor && !areEqualPeerDescriptors(this.localPeerDescriptor!, targetPeerDescriptor)) {
+            if (!isCompatibleVersion(sourceVersion, version)) {
+                managedConnection.rejectHandshake(HandshakeError.UNSUPPORTED_VERSION)
```
Comment (harbu) [inline packages/dht/src/connection/websocket/WebsocketConnector.ts:319]: Duplication of 
```ts
if (!isCompatibleVersion(sourceVersion, version)) {
    managedConnection.rejectHandshake(HandshakeError.UNSUPPORTED_VERSION)
}
```
Move before (outside) if-else block?

**Review by harbu — CHANGES_REQUESTED** [review]

**Review by teogeb — COMMENTED** [review]
We can remove some fallbacks in the future when all nodes send version info? Could add TODO comments (and link to Linear issue) in those places.

At that point, do we still need both request and response side handling?

**Review by teogeb — APPROVED** [review]

File: packages/dht/src/connection/websocket/WebsocketConnector.ts (hunk 3 above)
Comment (juslesan) [reply packages/dht/src/connection/websocket/WebsocketConnector.ts:319]: Yeah would be good to not have to do duplication here but it is necessary in this case

**Pull request merged**